### PR TITLE
TM-4276 AI Validation

### PR DESCRIPTION
**Pull request by elizabeth-jimenez**
Dual Merge: 
- [API PR](https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/803)

### CODE REVIEW

File: src/actions/agendaItemMaintenancePane.js [hunk 1]
```diff
@@ -47,20 +47,78 @@ export function aiCreate(panel, legs, personId, ef) {
...
+}
+export function validateAIIsLoading(bool) {
+  return {
+    type: 'VALIDATE_AI_IS_LOADING',
+    isLoading: bool,
+  };
+}
+export function validateAISuccess(data) {
+  /* eslint-disable no-console */
+  console.log('🐥🐥🐥🐥🐥🐥🐥🐥🐥🐥️');
```
Comment (elizabeth-jimenez) [inline src/actions/agendaItemMaintenancePane.js:84]: 🐤

File: src/sass/_agendaItemMaintenance.scss [hunk 2]
```diff
@@ -609,3 +609,9 @@
     display: flex;
   }
 }
+
+.sophie-error {
```
Comment (elizabeth-jimenez) [inline src/sass/_agendaItemMaintenance.scss:613]: 🙀

File: src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx [hunk 3]
```diff
@@ -128,6 +135,39 @@ const AgendaItemMaintenancePane = (props) => {
     selectedStatus,
     selectedPanelCat]);
 
+  useEffect(() => {
```
Comment (elizabeth-jimenez) [inline src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:138]: styling the different states of this button took half a day 😒

Comment (elizabeth-jimenez) [reply src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:138]: but it's hot tho 💅

File: src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx [hunk 4]
```diff
@@ -1,15 +1,15 @@
 import { useEffect, useRef, useState } from 'react';
 import FontAwesome from 'react-fontawesome';
-import { useDispatch } from 'react-redux';
+import { useDispatch, useSelector } from 'react-redux';
```
Comment (M00staff) [inline src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx:3]: i really can't stress enough, how much better Redux is with hooks

File: src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
 import { Tooltip } from 'react-tippy';
 import { withRouter } from 'react-router';
 import InteractiveElement from 'Components/InteractiveElement';
-import { drop, filter, find, get, has, isEmpty, sample } from 'lodash';
+import { drop, filter, find, get, has, isEmpty, isEqual, sample } from 'lodash';
```
Comment (M00staff) [inline src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx:7]: 👍 This is one of the few lodash functions that's actually worth using.

though - writing a vanilla solution and adding it to the project's constants would be something interesting and fun to play with on the side 🤔

File: src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx [hunk 6]
```diff
@@ -118,6 +125,19 @@ const AgendaItemMaintenanceContainer = (props) => {
     updateResearchPaneTab(RemarksGlossaryTabID);
   };
 
+  useEffect(() => {
+    if ((prevLegs && legs && !isEqual(prevLegs, legs)) ||
+      (prevMaintenanceInfo && maintenanceInfo && !isEqual(prevMaintenanceInfo, maintenanceInfo))) {
```
Comment (M00staff) [inline src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx:130]: So, I'm just noticing it now - but that `usePrevious` hook is pretty slick and an interesting use of `useRef`

I've only ever seen it for DOM manipulation

Comment (M00staff) [reply src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx:130]: you used to have to do this with a lifecycle hook IIRC

File: src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx [hunk 7]
```diff
@@ -128,6 +134,38 @@ const AgendaItemMaintenancePane = (props) => {
...
+
+    if (AIvalidationIsLoading) {
+      buttonMetadata.classNames = 'save-ai-btn button-tiny-loading-spinner min-width-155';
+      buttonMetadata.clickFunction = () => {};
+      buttonMetadata.disabled = true;
+      buttonMetadata.text = 'Validating AI';
+      buttonMetadata.children = (<span className="tiny-loading-spinner" />);
+    }
+
+    setValidationButton(buttonMetadata);
```
Comment (M00staff) [inline src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:162]: this is a VERY slick pattern, I'm kinda surprised it works and I don't think i've ever seen something like it

kudos! 🌟

File: src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx [hunk 8]
```diff
@@ -187,89 +230,106 @@ const AgendaItemMaintenancePane = (props) => {
...
-                  </select>
+                  <label className="select-label" htmlFor="ai-maintenance-status">Status:</label>
+                  <div className="error-message-wrapper">
+                    <div className="validation-error-message-label validation-error-message">
+                      {AIvalidation?.status?.errorMessage}
+                    </div>
+                    <select
+                      className={`aim-select ${AIvalidation?.status?.valid ? '' : 'validation-error-border'}`}
+                      id="ai-maintenance-status"
+                      onChange={(e) => setStatus(get(e, 'target.value'))}
```
Comment (M00staff) [inline src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:241]:
```suggestion
                      onChange={(e) => setStatus(e.target.value)}
```

File: src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx [hunk 9]
```diff
@@ -187,89 +230,106 @@ const AgendaItemMaintenancePane = (props) => {
...
-                  </select>
+                  <label className="select-label" htmlFor="ai-maintenance-report-category">Report Category:</label>
+                  <div className="error-message-wrapper">
+                    <div className="validation-error-message-label validation-error-message">
+                      {AIvalidation?.reportCategory?.errorMessage}
+                    </div>
+                    <select
+                      className={`aim-select ${AIvalidation?.reportCategory?.valid ? '' : 'validation-error-border'}`}
+                      id="ai-maintenance-category"
+                      onChange={(e) => setPanelCat(get(e, 'target.value'))}
```
Comment (M00staff) [inline src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:268]:
```suggestion
                      onChange={(e) => setPanelCat(e.target.value)}
```

File: src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx [hunk 10]
```diff
@@ -187,89 +230,106 @@ const AgendaItemMaintenancePane = (props) => {
...
+                  <label className="select-label" htmlFor="ai-maintenance-date">Panel Date:</label>
+                  <div className="error-message-wrapper">
+                    <div className="validation-error-message-label validation-error-message">
+                      {AIvalidation?.panelDate?.errorMessage}
+                    </div>
+                    <div>
+                      <select
+                        className={`aim-select-small ${AIvalidation?.panelDate?.valid ? '' : 'validation-error-border'}`}
+                        id="ai-maintenance-status"
+                        onChange={(e) => setDate(get(e, 'target.value'), true)}
```
Comment (M00staff) [inline src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:296]:
```suggestion
                        onChange={(e) => setDate(e.target.value, true)}
```

File: src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx [hunk 11]
```diff
@@ -187,89 +230,106 @@ const AgendaItemMaintenancePane = (props) => {
...
+                            >
+                              {get(a, 'pmt_code')} - {formatDate(get(a, 'pmd_dttm'))}
+                            </option>
+                          ))
+                        }
+                      </select>
+                      <select
+                        className={`aim-select-small ${AIvalidation?.panelDate?.valid ? '' : 'validation-error-border'}`}
+                        id="ai-maintenance-status"
+                        onChange={(e) => setDate(get(e, 'target.value'), false)}
```
Comment (M00staff) [inline src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:315]:
```suggestion
                        onChange={(e) => setDate(e.target.value, false)}
```

File: src/actions/agendaItemMaintenancePane.js [hunk 12]
```diff
@@ -47,20 +47,73 @@ export function aiCreate(panel, legs, personId, ef) {
...
+    dispatch(validateAIIsLoading(true));
+    dispatch(validateAIHasErrored(false));
+    api()
+      .post('/fsbid/agenda/agenda_item/validate/', {
+        ...ef,
+        personId,
+        ...panel,
+        agendaLegs: legs,
+      }, {
+        cancelToken: new CancelToken((c) => {
```
Comment (M00staff) [inline src/actions/agendaItemMaintenancePane.js:101]: this is a nifty axios feature!

File: src/reducers/agendaItemMaintenancePane/agendaItemMaintenancePane.js [hunk 13]
```diff
@@ -1,20 +1,20 @@
...
+export function aiCreateIsLoading(state = false, action) {
   switch (action.type) {
     case 'AI_CREATE_IS_LOADING':
       return action.isLoading;
     default:
       return state;
   }
 }
-export function aiCreateSuccess(state = [], action) {
+export function ai(state = [], action) {
```
Comment (M00staff) [inline src/reducers/agendaItemMaintenancePane/agendaItemMaintenancePane.js:17]: if we change this here - shouldn't we also change these lines, in the reducer?

or is it the `type` that its checking against?

image

Comment (M00staff) [reply src/reducers/agendaItemMaintenancePane/agendaItemMaintenancePane.js:17]: actually, i think it is the `type` that matters 🤔

**Review by M00staff — APPROVED** [review]
1 open question (that i think i answered myself)
a few 1 click suggested changes (that i don't even think were yours - just a part of the diff, you can ignore if you like)

nothing really holding this up - FYI, it looks AMAZING, and the little spinner on the save button was a VERY nice touch 🌟

File: src/reducers/agendaItemMaintenancePane/agendaItemMaintenancePane.js (hunk 13 above)
Comment (elizabeth-jimenez) [reply src/reducers/agendaItemMaintenancePane/agendaItemMaintenancePane.js:17]: Yea 😁

File: src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx [hunk 14]
```diff
@@ -169,8 +207,13 @@ const AgendaItemMaintenancePane = (props) => {
         <>
           <div className="back-save-btns-container">
             <BackButton />
-            <button className="save-ai-btn" onClick={saveAI} disabled={isReadOnly}>
-              Save Agenda Item
+            <button
+              className={validationButton?.classNames}
+              onClick={validationButton.clickFunction}
```
[the remark this answers is not in the document]
Comment (elizabeth-jimenez) [reply src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:212]: i think this one gave me an error, when i tried it

File: src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx (hunk 5 above)
Comment (elizabeth-jimenez) [reply src/Components/Agenda/AgendaItemMaintenanceContainer/AgendaItemMaintenanceContainer.jsx:7]: 🤔 yea, why not - we'll slowly but surely get rid of lodash lol

File: src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (M00staff) [reply src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:212]: i changed it locally and it worked fine - but i only clicked around once 🤷

[the remark this answers is not in the document]
Comment (elizabeth-jimenez) [reply src/Components/Agenda/AgendaItemMaintenancePane/AgendaItemMaintenancePane.jsx:212]: 🤔

**Comment by elizabeth-jimenez** [comment]
> 1 open question (that i think i answered myself) a few 1 click suggested changes (that i don't even think were yours - just a part of the diff, you can ignore if you like)
> 
> nothing really holding this up - FYI, it looks AMAZING, and the little spinner on the save button was a VERY nice touch 🌟

Since I branched off of this PR for my other one, I'll implement the lodash suggestions in that one 😁

**Pull request merged**